### PR TITLE
Refactor FUSE related codes for decoupling

### PR DESCRIPTION
**Pull request by JiamingMai**
Refactor FUSE related codes for decoupling:
1. Extract FuseStreamFactory to an interface.
2. Provide SimpleFuseStreamFactory which is an implementation of FuseStreamFactory.
3. Provide preprocess method for processing the instance of AlluxioJniFuseFileSystem before doing FUSE mount.

### CODE REVIEW

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java [hunk 1]
```diff
@@ -210,34 +226,70 @@ public static FuseUmountable launchFuse(FileSystemContext fsContext, FileSystem
       }
     }
 
-    final boolean debugEnabled = conf.getBoolean(PropertyKey.FUSE_DEBUG_ENABLED);
     if (conf.getBoolean(PropertyKey.FUSE_JNIFUSE_ENABLED)) {
       final AlluxioJniFuseFileSystem fuseFs
           = new AlluxioJniFuseFileSystem(fsContext, fs, fuseOptions);
+      return fuseFs;
+    } else {
```
Comment (elega) [inline dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:233]: as a side question, should we remove jnr fuse in alluxio 300? @LuQQiu

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/SimpleFuseStreamFactory.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        if (mPositionReadEnabled) {
+          return FusePositionReader.create(mFileSystem, mLockManager, uri);
+        }
+        return FuseFileInStream.create(mFileSystem, mLockManager, uri);
+      case O_WRONLY:
+        return FuseFileOutStream.create(mFileSystem, mAuthPolicy, mLockManager, uri, flags, mode);
+      default:
+        return FuseFileInOrOutStream.create(mFileSystem, mAuthPolicy, mLockManager,
+            uri, flags, mode);
+        /*
```
Comment (elega) [inline dora/integration/fuse/src/main/java/alluxio/fuse/file/SimpleFuseStreamFactory.java:63]: do we still need this?

File: dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java [hunk 3]
```diff
@@ -115,6 +116,7 @@ public void overwriteEmpty() {
     }, true);
   }
 
+  @Ignore("RandomAccessFuseFileStream allows to overwrite the existing file")
```
Comment (elega) [inline dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java:119]: why do we ignore these? assuming you are implementing RandomAccessFuseFileStream somewhere else but it should not break existing tests

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java [hunk 4]
```diff
@@ -210,34 +226,70 @@ public static FuseUmountable launchFuse(FileSystemContext fsContext, FileSystem
...
+      return fuseFs;
+    }
+  }
 
+  /**
+   * Preprocess the instance of {@link AlluxioJniFuseFileSystem} before mounting.
+   *
+   * @param jniFuseFileSystem the instance of {@link AlluxioJniFuseFileSystem}
+   */
+  public void preprocess(AlluxioJniFuseFileSystem jniFuseFileSystem) {
```
Comment (elega) [inline dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:244]: this naming is a bit unnatural to me. Can we rename this to something like setupFuseFileSystem?

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java (hunk 1 above)
Comment (huanghua78) [reply dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:233]: I thought @LuQQiu had a plan/task to remove the JNR implementation totally from Alluxio code base.

If this is not planned, maybe we can do it in this task.

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java (hunk 4 above)
Comment (huanghua78) [reply dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:244]: So, this function is empty and can be overrided to add special handlings?

File: dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java (hunk 3 above)
Comment (huanghua78) [reply dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java:119]: If "RandomAccessFuseFileStream allows to overwrite the existing file", then we should improve the test to verify this feature.

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java (hunk 1 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:233]: I think JNR can be removed, but maybe we can remove JNR in another PR.

File: dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java (hunk 3 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/test/java/alluxio/fuse/ufs/FuseFileSystemDataTest.java:119]: Oh, I forget to remove the @Ignore annotation. Let me update the code later.

File: dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java (hunk 4 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:244]: > So, this function is empty and can be overrided to add special handlings?
Actually, this method can be regarded as a hook method, and we can override it in the subclass of AlluxioFuse (e.g. the EE class).

Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuse.java:244]: > this naming is a bit unnatural to me. Can we rename this to something like setupFuseFileSystem?

Sure, `setupFuseFileSystem` is better. The method is for the action before startup.

File: dora/integration/fuse/src/main/java/alluxio/fuse/file/SimpleFuseStreamFactory.java (hunk 2 above)
Comment (JiamingMai) [reply dora/integration/fuse/src/main/java/alluxio/fuse/file/SimpleFuseStreamFactory.java:74]: We still need this to support rw mode although `FuseFileInOrOutStream` doesn't support "reading while writing/truncating".

**Review by elega — APPROVED** [review]
LGTM

**Comment by JiamingMai** [comment]
alluxio-bot, merge this please

**Pull request merged**